File: pra-module-1/cpp/tests/test_domain.cpp
#include "mini_catch.hpp"
#include "pinjamlab/equipment.hpp"
#include "pinjamlab/laptop.hpp"
#include "pinjamlab/bundle.hpp"
#include "pinjamlab/member.hpp"
#include "pinjamlab/loan.hpp"
#include <vector>
#include <memory>
#include <numeric>

using namespace pinjamlab;

// Test concrete subclass for original artifact verification
class Projector : public Equipment {
private:
    int lumens_;
public:
    Projector(AssetTag tag, std::string model, int lumens = 3600)
        : Equipment(std::move(tag), std::move(model)), lumens_(lumens) {}

    [[nodiscard]] int dailyFine() const override {
        return 35000; // Rp 35,000 / day
    }

    [[nodiscard]] std::string categoryName() const override {
        return "Projector";
    }
};

TEST_CASE("Polymorphism: Dynamic dispatch across Equipment hierarchy") {
    std::vector<std::unique_ptr<Equipment>> inventory;
    inventory.push_back(std::make_unique<Laptop>(AssetTag("LP-001"), "ThinkPad T14", 16));
    inventory.push_back(std::make_unique<Projector>(AssetTag("PJ-001"), "Epson EB-X51", 3800));

    REQUIRE(inventory[0]->dailyFine() == 50000);
    REQUIRE(inventory[1]->dailyFine() == 35000);

    int totalDailyFine = 0;
    for (const auto& item : inventory) {
        totalDailyFine += item->dailyFine();
    }
    REQUIRE(totalDailyFine == 85000);
}

TEST_CASE("Equipment: State transitions and exception guards") {
    Laptop laptop(AssetTag("LP-001"), "ThinkPad T14");
    REQUIRE(laptop.getState() == EquipmentState::AVAILABLE);

    laptop.markLoaned();
    REQUIRE(laptop.getState() == EquipmentState::LOANED);

    REQUIRE_THROWS_AS(laptop.markLoaned(), std::runtime_error);

    laptop.markReturned();
    REQUIRE(laptop.getState() == EquipmentState::AVAILABLE);
}

TEST_CASE("Member: Encapsulation and quota limits") {
    Member student("M-101", "Budi Santoso", Role::STUDENT);
    REQUIRE(student.getMaxLoans() == 2);
    REQUIRE(student.getActiveLoans() == 0);
    REQUIRE(student.canBorrow() == true);

    student.recordLoan();
    student.recordLoan();
    REQUIRE(student.getActiveLoans() == 2);
    REQUIRE(student.canBorrow() == false);

    REQUIRE_THROWS_AS(student.recordLoan(), std::runtime_error);

    student.recordReturn();
    REQUIRE(student.getActiveLoans() == 1);
    REQUIRE(student.canBorrow() == true);
}

TEST_CASE("Loan: Overdue fine calculation and ordering") {
    auto laptop = std::make_shared<Laptop>(AssetTag("LP-001"), "ThinkPad T14");
    auto student = std::make_shared<Member>("M-101", "Budi Santoso", Role::STUDENT);

    Loan loan1("L-001", laptop, student, 7, 3);
    // 3 overdue days * 50,000 = 150,000
    REQUIRE(loan1.calculateFine() == 150000);

    auto projector = std::make_shared<Projector>(AssetTag("PJ-001"), "Epson EB-X51");
    Loan loan2("L-002", projector, student, 3, 2);
    // 2 overdue days * 35,000 = 70,000
    REQUIRE(loan2.calculateFine() == 70000);

    REQUIRE(loan1 < loan2);
}

TEST_CASE("Composition over inheritance: EquipmentBundle") {
    auto laptop = std::make_shared<Laptop>(AssetTag("LP-001"), "ThinkPad T14");
    auto projector = std::make_shared<Projector>(AssetTag("PJ-001"), "Epson EB-X51");

    EquipmentBundle bundle(AssetTag("BN-001"), "Multimedia Presentation Kit");
    bundle.addItem(laptop);
    bundle.addItem(projector);

    REQUIRE(bundle.categoryName() == "EquipmentBundle");
    REQUIRE(bundle.getItems().size() == 2);

    // Sum = 50,000 + 35,000 = 85,000 -> 10% discount = 76,500
    REQUIRE(bundle.dailyFine() == 76500);
}

int main() {
    return test_runner::runAll();
}

File: pra-module-1/py/tests/test_domain.py
"""Integration and domain tests for PinjamLab hierarchy and polymorphic loans."""
import pytest
from pinjamlab.asset_tag import AssetTag
from pinjamlab.equipment import Equipment, EquipmentState
from pinjamlab.laptop import Laptop
from pinjamlab.bundle import EquipmentBundle
from pinjamlab.member import Member, Role
from pinjamlab.loan import Loan


class Projector(Equipment):
    """Custom equipment subclass for test and original artifact verification."""

    def __init__(self, tag: AssetTag, model: str, lumens: int = 3600) -> None:
        super().__init__(tag, model)
        self.lumens = lumens

    def daily_fine(self) -> int:
        return 35000

    def category_name(self) -> str:
        return "Projector"


def test_abstract_class_cannot_be_instantiated():
    with pytest.raises(TypeError):
        Equipment(AssetTag("XX-001"), "Generic Tool")  # type: ignore


def test_polymorphic_daily_fine_and_dispatch():
    laptop = Laptop(AssetTag("LP-001"), "ThinkPad T14", ram_gb=32)
    projector = Projector(AssetTag("PJ-001"), "Epson EB-X51", lumens=3800)

    assert laptop.daily_fine() == 50000
    assert projector.daily_fine() == 35000

    items: list[Equipment] = [laptop, projector]
    fines = [item.daily_fine() for item in items]
    assert fines == [50000, 35000]
    assert sum(fines) == 85000


def test_equipment_state_transitions():
    laptop = Laptop(AssetTag("LP-001"), "ThinkPad T14")
    assert laptop.state == EquipmentState.AVAILABLE

    laptop.mark_loaned()
    assert laptop.state == EquipmentState.LOANED

    with pytest.raises(RuntimeError):
        laptop.mark_loaned()  # Already loaned

    laptop.mark_returned()
    assert laptop.state == EquipmentState.AVAILABLE


def test_member_quota_encapsulation():
    student = Member("M-101", "Budi Santoso", Role.STUDENT)
    assert student.max_loans == 2
    assert student.active_loans == 0
    assert student.can_borrow() is True

    student.record_loan()
    student.record_loan()
    assert student.active_loans == 2
    assert student.can_borrow() is False

    with pytest.raises(RuntimeError):
        student.record_loan()  # Exceeds quota

    student.record_return()
    assert student.can_borrow() is True


def test_loan_fine_calculation_and_ordering():
    laptop = Laptop(AssetTag("LP-001"), "ThinkPad T14")
    student = Member("M-101", "Budi Santoso", Role.STUDENT)

    loan1 = Loan("L-001", laptop, student, days_loaned=7, overdue_days=3)
    # Fine = 3 days * 50,000 = 150,000
    assert loan1.calculate_fine() == 150000

    projector = Projector(AssetTag("PJ-001"), "Epson EB-X51")
    loan2 = Loan("L-002", projector, student, days_loaned=3, overdue_days=2)
    # Fine = 2 days * 35,000 = 70,000
    assert loan2.calculate_fine() == 70000

    # Total ordering by loan_id
    assert loan1 < loan2
    loans = [loan2, loan1]
    loans.sort()
    assert loans == [loan1, loan2]


def test_composition_over_inheritance_bundle():
    laptop = Laptop(AssetTag("LP-001"), "ThinkPad T14")
    projector = Projector(AssetTag("PJ-001"), "Epson EB-X51")

    bundle = EquipmentBundle(AssetTag("BN-001"), "Presentation Kit", [laptop, projector])
    assert bundle.category_name() == "EquipmentBundle"
    assert len(bundle.items) == 2
    # 50,000 + 35,000 = 85,000 -> 10% discount = 76,500
    assert bundle.daily_fine() == 76500
